### PR TITLE
feature(plugin): rewrite upstream info with plugin `proxy-rewrite`

### LINKED ISSUE #569: change: implement a new plugin, to modify scheme, uri, host and enable_websocket.
we can change `scheme`, `uri`, `host` and `enable_websocket` when we create a new upstream object now, this is not appropriate.

They should not be upstream attributes and should exist independently. We should also allow different uri values to be set for the same upstream.

在当前方案中，我们可以在创建 upstream 对象时指定 `scheme`, `uri`, `host` and `enable_websocket` 等值，这是不合适的。

这些不应当是 upstream 的属性。比如同一个上游，不同请求下应当允许被设置成不同 uri 。

### PR DESCRIPTION
**Pull request by shuaijinchao**
This plugin is used to separate `uri`, `host`, `scheme` and `enable_websocket` work in the `upstream` module. Through this plugin, the simplicity and independence of the `upstream` module can be maintained.
此插件用于分离`upstream`模块中的`uri`、`host`、`scheme`和`enable_websocket`功能。让`upstream`模块可以保持简洁和独立性。

### Issues resolved

Fix #569

### CODE REVIEW

File: lua/apisix/plugins/proxy-rewrite.lua [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+        end
+    end
+
+    if conf.enable_websocket then
+        ctx.var["upstream_upgrade"]    = ctx.var["http_upgrade"]
+        ctx.var["upstream_connection"] = ctx.var["http_connection"]
+    end
+
+    -- rewrite nodes for upstream config ,priority is higher than upstream nodes.
+    if conf.host then
```
Comment (membphis) [inline lua/apisix/plugins/proxy-rewrite.lua]: I guess this `host` should be the host domain in HTTP header, all right?

**Comment by membphis** [comment]
https://travis-ci.org/iresty/apisix/builds/589504274#L810

please fix it first.

File: lua/apisix/plugins/proxy-rewrite.lua (hunk 1 above)
Comment (shuaijinchao) [reply lua/apisix/plugins/proxy-rewrite.lua]: OK, my understanding is the host address.

Comment (membphis) [reply lua/apisix/plugins/proxy-rewrite.lua]: so we need to delete this code from line 70 to 74.

File: lua/apisix/plugins/proxy-rewrite.lua [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    for name, _ in pairs(upstream_vars) do
+        core.table.insert(upstream_names, name)
+    end
+
+function _M.rewrite(conf, ctx)
+    local ngx_var = ngx.var
+
+    for _, name in ipairs(upstream_names) do
+        if conf[name] then
+            ngx_var[upstream_vars[name]] = conf[name]
```
Comment (membphis) [inline lua/apisix/plugins/proxy-rewrite.lua]: `ctx.var[upstream_vars[name]] = conf[name]`

it'll update `ngx.var` automatically, source code: https://github.com/iresty/apisix/blob/master/lua/apisix/core/ctx.lua#L65

File: doc/plugins/proxy-rewrite.md [hunk 3]
```diff
@@ -0,0 +1,75 @@
+[中文](proxy-rewrite-cn.md)
+# proxy-rewrite
+
+upstream proxy info rewrite plugin.
+
+### Parameters
+|Name    |Must|Description|
```
Comment (membphis) [inline doc/plugins/proxy-rewrite.md]: Must -> Required

File: doc/plugins/proxy-rewrite-cn.md [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+        "type": "roundrobin",
+        "nodes": {
+            "127.0.0.1:80": 1
+        }
+    }
+}'
+```
+
+#### 测试插件
+基于上述配置进行测试，测试前在 Nginx 配置文件 access 日志输出配置中补充 `$upstream_scheme $upstream_addr $upstream_uri` 变量。
```
Comment (moonming) [inline doc/plugins/proxy-rewrite-cn.md]: APISIX not support to change the format of access log, so this method is not good enough. 
I think check the asscees log of upstream server is better.

File: doc/plugins/proxy-rewrite.md [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+|-------         |-----|------|
+|scheme          |No| Upstream new `schema` forwarding protocol,options can be `http` or `https`,default `http`.|
+|uri             |No| Upstream new `uri` forwarding address.|
+|host            |No| Upstream new `host` forwarding address, can be `192.168.80.128:8080` or `192.168.80.128` format, not set up port default `80`, priority over `upstream.nodes` configuration. |
+|enable_websocket|No| enable `websocket`(boolean), default disable.|
```
Comment (moonming) [inline doc/plugins/proxy-rewrite.md]: default value should be `false`, not `disable`

File: lua/apisix/plugins/proxy-rewrite.lua [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+    if conf.enable_websocket then
+        ctx.var["upstream_upgrade"]    = ctx.var["http_upgrade"]
+        ctx.var["upstream_connection"] = ctx.var["http_connection"]
+    end
+
+    -- rewrite nodes for upstream config ,priority is higher than upstream nodes.
+    if conf.host then
+        local tmp_nodes = {}
+        tmp_nodes[conf.host] = 100
```
Comment (moonming) [inline lua/apisix/plugins/proxy-rewrite.lua]: the magic number `100` is for what?

**Comment by shuaijinchao** [comment]
OK, I will fix one by one.

**Comment by moonming** [comment]
@shuaijinchao please press the `resolve conversation` button after you fix one problem, so that's will save time for us. GitHub's code review is powerful.
thx

File: doc/plugins/proxy-rewrite-cn.md [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+# proxy-rewrite
+
+上游代理信息重写插件。
+
+#### 配置参数
+|名字    |可选|说明|
+|-------         |-----|------|
+|scheme          |可选| 转发到上游的新`schema` 协议，可以是`http`或`https`，默认`http`协议|
+|uri             |可选| 转发到上游的新`uri` 地址|
+|host            |可选| 转发到上游的新`host` 地址，格式可以为`192.168.80.128:8080`或`192.168.80.128`如果未设置端口将默认设置为`80`，此配置优先级将高于`upstream.nodes` |
```
[the remark this answers is not in the document]
Comment (shuaijinchao) [reply doc/plugins/proxy-rewrite-cn.md]: headers rewriting is currently not supported, the purpose of this plugin is to migrate the `uri`, `host`, `scheme` and `enable_websocket` feature in the upstream module. this suggestion is good, I will add this feature later.

File: t/lib/server.lua [hunk 8]
```diff
@@ -22,6 +23,13 @@ function _M.limit_conn()
     ngx.say("hello world")
 end
 
+function _M.plugin_proxy_rewrite()
+    local req_info = {}
+    req_info['uri'] = ngx.var.uri
+    req_info['host'] = ngx.var.host
+    req_info['scheme'] = ngx.var.scheme
+    ngx.say(json_encode(req_info))
```
Comment (membphis) [inline t/lib/server.lua]: The result of json serialization is unstable, such as the order of the different keys.

So it's safer to print the value of each item directly.

```lua
ngx.say("uri: ", ngx.var.uri)
ngx.say("host: ", ngx.var.host)
...
```

File: doc/plugins/proxy-rewrite-cn.md [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+|-------         |-----|------|
+|scheme          |可选| 转发到上游的新`schema` 协议，可以是`http`或`https`，默认`http`协议|
+|uri             |可选| 转发到上游的新`uri` 地址|
+|host            |可选| 转发到上游的新`host` 地址，例如：`iresty.com` |
+|enable_websocket|可选| 是否启用`websocket`（布尔值），默认不启用|
+
+### 示例
+
+#### 开启插件
+下面是一个示例，在指定的 route 上开启了 proxy rewrite 插件:
```
Comment (membphis) [inline doc/plugins/proxy-rewrite-cn.md]: `proxy rewrite` is a special noun, so it's best to wrap it around with \`, example: \`xxxx\`

File: doc/plugins/proxy-rewrite-cn.md [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+
+发送请求，查看上游服务`access.log`，如果输出信息与配置一致：
+```
+127.0.0.1 - [26/Sep/2019:10:52:20 +0800] iresty.com GET /test/home.html HTTP/1.1 200 38 - curl/7.29.0 - 0.000 199 107
+```
+
+即表示 proxy rewrite 插件生效了。
+
+#### 禁用插件
+当你想去掉 proxy rewrite 插件的时候，很简单，在插件的配置中把对应的 json 配置删除即可，无须重启服务，即刻生效：
```
Comment (membphis) [inline doc/plugins/proxy-rewrite-cn.md]: ditto

File: doc/plugins/proxy-rewrite-cn.md [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+```shell
+curl -X GET http://127.0.0.1:9080/test/index.html
+```
+
+发送请求，查看上游服务`access.log`，如果输出信息与配置一致：
+```
+127.0.0.1 - [26/Sep/2019:10:52:20 +0800] iresty.com GET /test/home.html HTTP/1.1 200 38 - curl/7.29.0 - 0.000 199 107
+```
+
+即表示 proxy rewrite 插件生效了。
```
Comment (membphis) [inline doc/plugins/proxy-rewrite-cn.md]: ditto

File: doc/plugins/proxy-rewrite-cn.md [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+    "upstream": {
+        "type": "roundrobin",
+        "nodes": {
+            "127.0.0.1:80": 1
+        }
+    }
+}'
+```
+
+现在就已经移除了 proxy rewrite 插件了。其他插件的开启和移除也是同样的方法。
```
Comment (membphis) [inline doc/plugins/proxy-rewrite-cn.md]: Please fix similar issues in the English documentation.

File: lua/apisix/plugins/proxy-rewrite.lua [hunk 13]
```diff
@@ -0,0 +1,69 @@
+local core        = require("apisix.core")
+local plugin_name = "proxy-rewrite"
+local pairs       = pairs
+local ipairs      = ipairs
+
+local schema = {
+    type = "object",
+    properties = {
+        uri = {
+            type = "string"
```
Comment (membphis) [inline lua/apisix/plugins/proxy-rewrite.lua]: add more meaningful restrictions: `minLengh`, `maxLengh`

File: lua/apisix/plugins/proxy-rewrite.lua [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+        },
+        host = {
+            type = "string"
```
Comment (membphis) [inline lua/apisix/plugins/proxy-rewrite.lua]: simple domain name regular, take a look at: https://github.com/iresty/apisix/blob/master/lua/apisix/schema_def.lua#L258

File: lua/apisix/plugins/proxy-rewrite.lua [hunk 15]
```diff
@@ -0,0 +1,69 @@
+local core        = require("apisix.core")
+local plugin_name = "proxy-rewrite"
+local pairs       = pairs
+local ipairs      = ipairs
+
+local schema = {
+    type = "object",
+    properties = {
```
Comment (membphis) [inline lua/apisix/plugins/proxy-rewrite.lua]: http://json-schema.org/latest/json-schema-validation.html#rfc.section.6.5.2

Make sure the user has at least one of them enabled.

File: t/plugin/proxy-rewrite.t [hunk 16]
```diff
@@ -0,0 +1,235 @@
...
+    }
+--- request
+GET /t
+--- response_body
+passed
+--- no_error_log
+[error]
+
+
+=== TEST 6: set route(id: 1)
```
Comment (moonming) [inline t/plugin/proxy-rewrite.t]: need add more test cases to cover `uri`,`scheme`,`host`,`enable_websocket`

Comment (membphis) [reply t/plugin/proxy-rewrite.t]: I'll do this later after this PR merged.

**Comment by membphis** [comment]
@shuaijinchao merged, many thx

**Pull request merged**